### PR TITLE
Introduce nightly feature

### PR DESCRIPTION
**Pull request by Aceeri**
Adds a new feature flag `nightly` for unstable, but useful features.

Also adds first feature using the flag which allows clearer panic messages when not adding/registering resources/components.

Examples of new error message when enabled:
```
thread 'main' panicked at 'Component `Pos` was unregistered. Did you forget to register the component with `World::register::<Pos>()`', src\libcore\option.rs:839:4
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: process didn't exit successfully: `target\debug\examples\cluster_bomb.exe` (exit code: 101)
```
```
thread 'main' panicked at 'No resource `specs::saveload::U64MarkerAllocator` exists in the world. Did you forget to add the resource with `World::add_resource::<specs::saveload::U64MarkerAllocator>(...)`', src\libcore\option.rs:839:4
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: process didn't exit successfully: `target\debug\examples\saveload.exe` (exit code: 101)
```

### CODE REVIEW

File: src/world/mod.rs [hunk 1]
```diff
@@ -16,11 +16,49 @@ mod lazy;
...
 const COMPONENT_NOT_REGISTERED: &str = "No component with the given id. Did you forget to register \
                                         the component with `World::register::<ComponentName>()`?";
+#[cfg(not(feature = "nightly"))]
 const RESOURCE_NOT_ADDED: &str = "No resource with the given id. Did you forget to add \
                                   the resource with `World::add_resource(resource)`?";
 
+#[cfg(feature = "nightly")]
+fn nightly_component_message<T>(id: usize) -> String {
+    let type_name = unsafe { ::std::intrinsics::type_name::<T>() };
+    let (message, function) = if id == 0 {
```
Comment (torkleyy) [inline src/world/mod.rs]: Hm, this formatting looks weird

File: src/world/mod.rs [hunk 2]
```diff
@@ -16,11 +16,49 @@ mod lazy;
 #[cfg(test)]
 mod tests;
 
+#[cfg(not(feature = "nightly"))]
```
Comment (torkleyy) [inline src/world/mod.rs]: Could we add something like `"Note: enable nightly feature to get the component type printed out"`?

File: src/world/mod.rs (hunk 1 above)
Comment (Aceeri) [reply src/world/mod.rs]: Not sure how I could make it better without making the lines super long though. `cargo +nightly fmt` doesn't seem to do anything to it either.

**Comment by Aceeri** [comment]
@torkleyy This should be ready I believe

File: Cargo.toml [hunk 3]
```diff
@@ -36,9 +36,10 @@ rudy = { version = "0.1", optional = true }
 
 [features]
 common = ["futures"]
+nightly = []
 
 [package.metadata.docs.rs]
-features = ["common", "serde"]
+features = ["common", "serde", "nightly"]
```
Comment (torkleyy) [inline Cargo.toml]: Does docs.rs build with the nightly toolchain?

File: src/world/mod.rs [hunk 4]
```diff
@@ -16,11 +16,67 @@ mod lazy;
...
+/// Note: Enable nightly feature to get the component type
+/// printed instead of just a vague catch all message.
```
Comment (torkleyy) [inline src/world/mod.rs]: I'm not sure how this doc comment helps a user getting a panic? I thought we could include this note in the string.

File: src/world/mod.rs [hunk 5]
```diff
@@ -311,7 +369,13 @@ impl World {
         let entities = self.entities();
         let resource = self.res.try_fetch::<MaskedStorage<T>>(id);
 
-        Storage::new(entities, resource.expect(COMPONENT_NOT_REGISTERED))
+
+        #[cfg(not(feature = "nightly"))]
+        let expect_message = COMPONENT_NOT_REGISTERED;
+        #[cfg(feature = "nightly")]
+        let expect_message = &nightly_component_message::<T>(id);
```
Comment (WaDelma) [inline src/world/mod.rs]: I would move these messages to function and guard that with `cfg`:
````rust
#[cfg(not(feature = "nightly"))]
fn component_message<T>(id: usize) -> &'static str {
    COMPONENT_NOT_REGISTERED
}
#[cfg(feature = "nightly")]
fn component_message<T>(id: usize) -> String {
    // Create message based on type
}
````

**Comment by torkleyy** [comment]
Thank you!

bors r+

**Pull request merged**